### PR TITLE
Add FBA Channel type support

### PR DESCRIPTION
**Pull request by norkans7**

### CODE REVIEW

File: backends/rapidpro/channel.go [hunk 1]
```diff
@@ -139,21 +139,143 @@ const localTTL = 60 * time.Second
...
+       ch.config as config,
+       org.config as org_config,
+       org.is_anon as org_is_anon
+FROM
+       channels_channel ch
+       JOIN orgs_org org on ch.org_id = org.id
+WHERE
+       ch.address = $1 AND
+       ch.is_active = true AND
+       ch.org_id IS NOT NULL`
```
Comment (nicpottier) [inline backends/rapidpro/channel.go:195]: Don't think we need org_id is not null anymore..

File: backends/rapidpro/channel.go [hunk 2]
```diff
@@ -139,21 +139,143 @@ const localTTL = 60 * time.Second
...
+
+// ChannelForUUID attempts to look up the channel with the passed in UUID, returning it
```
Comment (nicpottier) [inline backends/rapidpro/channel.go:197]: update comment

File: backends/rapidpro/channel.go [hunk 3]
```diff
@@ -139,21 +139,143 @@ const localTTL = 60 * time.Second
...
+func loadChannelByAdrressFromDB(ctx context.Context, db *sqlx.DB, channelType courier.ChannelType, address courier.ChannelAddress) (*DBChannel, error) {
```
Comment (nicpottier) [inline backends/rapidpro/channel.go:198]: typo

File: handlers/facebookapp/facebookapp.go [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+
+		// this was wired successfully
+		status.SetStatus(courier.MsgWired)
+	}
+
+	return status, nil
+}
+
+// ReceiveVerify handles Facebook's webhook verification callback
+func (h *handler) DescribeURN(ctx context.Context, channel courier.Channel, urn urns.URN) (map[string]string, error) {
```
Comment (nicpottier) [inline handlers/facebookapp/facebookapp.go:616]: Comment wrong

File: handlers/forms.go [hunk 5]
```diff
@@ -96,3 +93,12 @@ func DecodeAndValidateXML(envelope interface{}, r *http.Request) error {
 
 	return nil
 }
+
+// ReadBody of a HTTP request up to limit bytes and make sure the Body is not consumed
+func ReadBody(r *http.Request, limit int64) ([]byte, error) {
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, limit))
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	defer r.Body.Close()
```
Comment (nicpottier) [inline handlers/forms.go:101]: This isn't doing anything. (NoopCloser does nothing on Close)

File: server.go [hunk 6]
```diff
@@ -324,7 +318,7 @@ func (s *server) channelHandleWrapper(handler ChannelHandler, handlerFunc Channe
 		}
 
 		// if no events were created we still want to log this to the channel, do so
-		if len(events) == 0 {
+		if len(events) == 0 && channel != nil {
```
Comment (nicpottier) [inline server.go:321]: in what scenario do we get here without a channel but with events?

Comment (norkans7) [reply server.go:321]: we use the URL `/c/fba/receive` for the channel type so we will figure out the channel in the handler(not here by UUID)

Comment (nicpottier) [reply server.go:321]: Why isn't the channel being returned by the FBA channel in `GetChannel`?

Comment (norkans7) [reply server.go:321]: Right that is not needed

Comment (norkans7) [reply server.go]: I remembered why we needed this we only add channel logs when we have a channel matched
https://github.com/nyaruka/courier/pull/311

**Pull request merged**